[10.0][ADD] HR Extra Hours Management

ecino opened the pull request:
This adds a new module for managing extra hours of employees in Odoo.

- Each employee can have control of his extra hours balance
- Managers can define attendance rules (like minimum break, due hours, maximum extra hours allowed)

ecino committed:
[ADD] hr_attendance_extra_hours module

pedrobaeza commented:
Please remove plurals in module name, follow guidelines and check Travis status.

ecino commented:
extra_hour doesn't mean anything, are you sure I should remove the plural? it's not related to a database object, rather to say "extra hours management module".

ecino commented:
The travis error are on another module tests: hr_holidays_validity_date. These tests are failing only on 1st of months because they create data conflicting with hr_holidays_demo! I will add a commit to fix it.

pedrobaeza commented:
The module name is a technical one, and the idea is to avoid typo problems with plurals. Check https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#file-naming where it's stated.

ecino commented:
What about `hr_holidays_compute_days`, `hr_public_holidays`, etc.. ?
I can't figure out another name that would represent what the module is doing.
`hr_attendance_` is the good prefix as it builds upon that.
It could be `hr_attendance_day` because we use such a new model in the module but this doesn't tell much... `hr_attendance_extra_hour` doesn't mean anything. `hr_attendance_balance` could be a candidate but I like it less.

I'm open to other propositions.

pedrobaeza commented:
See the exception also in the same link: if the Odoo module already contains plural, then it should be preserved.

pedrobaeza commented:
And there are some of them that are being renamed already (this rule is since version 8), but some of them come from before.

ecino commented:
Tests are fixed.

ecino commented:
@pedrobaeza I found a better name : `hr_attendance_management`

open-net-sarl commented:
@ecino I found a bug when trying to create an attendance day manually.
Traceback (most recent call last):
  File "/home/odoo/git/Odoo/odoo/odoo/http.py", line 641, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/odoo/git/Odoo/odoo/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/home/odoo/git/Odoo/odoo/odoo/http.py", line 333, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/odoo/git/Odoo/odoo/odoo/service/model.py", line 101, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/odoo/git/Odoo/odoo/odoo/http.py", line 326, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/odoo/git/Odoo/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/home/odoo/git/Odoo/odoo/odoo/http.py", line 506, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/git/Odoo/odoo/addons/web/controllers/main.py", line 892, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/odoo/git/Odoo/odoo/addons/web/controllers/main.py", line 884, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/odoo/git/Odoo/odoo/odoo/api.py", line 687, in call_kw
    return call_kw_model(method, model, args, kwargs)
  File "/home/odoo/git/Odoo/odoo/odoo/api.py", line 672, in call_kw_model
    result = method(recs, *args, **kwargs)
  File "/home/odoo/git/Odoo/odoo/odoo/models.py", line 1120, in default_get
    defaults[name] = field.default(self)
  File "/home/odoo/addons/hr_attendance_management/models/hr_attendance_day.py", line 36, in <lambda>
    default=lambda self: self.env.user.employee_ids[0].id)
  File "/home/odoo/git/Odoo/odoo/odoo/models.py", line 5236, in __getitem__
    return self._browse((self._ids[key],), self.env)
IndexError: tuple index out of range

ecino commented:
@open-net-sarl Thanks for pointing out the bug, I just fixed it.

open-net-sarl commented:
@ecino Thanks for the quick fix. It's working fine now.
But i noticed something strange. When the break is not long enough you modified the attendances.
You put a checkbox to notify that change. I don't think it's a good idea to modify the attendances.
I would advise to only modify the hr.attendance.days. 
Thanks,

ecino commented:
@open-net-sarl the attendances are never modified. it's another model `hr.attendance.break` which contains a break duration that is modified. Plus, this behaviour can be disabled if you remove the break rules settings (see readme file).

open-net-sarl commented:
@ecino I saw that a few minutes after posting the message.
Sorry for the comment.

open-net-sarl commented:
@ecino I have tested the module.
So far it's seems to work as expected.
There is nearly everything you would expect from that kind of module except the following thing:
- The break rule minimal total duration should be variable depending of the number of worked hours.
For example :
- If i have worked more than 6 hours i should have taken 30 minutes of break.
- If i have worked more than 8 hours i should have taken 45 minutes of break.
These rules should also depends of the employee and the day of the week no ?

Thanks a lot for this module it's a big improvement  from the last hr_attendance_calendar.

ecino commented:
@open-net-sarl Thanks for the suggestions!

- I added in the README/roadmap your suggestions for defining specific rules per employee tag and day of the week. I think employee tag is more relevant than employee only, so you could share some rules with more than one employee (or only one and you create a unique tag for one employee).
- You can already create different rules depending of the number of worked hours. See the below example of configuration which corresponds to the legal law in Switzerland (and is also set in the demo data of the module).

![image](https://user-images.githubusercontent.com/8435180/39689237-54c9a07a-51d6-11e8-9c9f-27c1b537053a.png)

open-net-sarl commented:
@ecino Thanks for the informations.
I tought the "From" and the "To" was the time slot when you need to take a break. Not the duration of the break.
I agree about the employee tag.

+1

ecino commented:
@open-net-sarl Thanks for your review. Can you approve the pull request when you think it's good for you?

open-net-sarl commented:
@ecino Of course.

ecino committed:
Fixes for hours recomputation

ecino committed:
Improve change day request wizard code

ecino commented:
@open-net-sarl I did slight adjuments. Do you approve it?

MeetKD commented on hr_attendance_management/__manifest__.py:
```diff
@@ -0,0 +1,71 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
this can be a short license header

MeetKD commented on hr_attendance_management/__manifest__.py:
```diff
@@ -0,0 +1,71 @@
+    ],
+    'version': '10.0.1.1.0',
+    'auto_install': False,
+    'website': 'https://www.compassion.ch',
```
website should be https://github.com/OCA/hr

MeetKD commented on hr_attendance_management/__manifest__.py:
```diff
@@ -0,0 +1,71 @@
+    'website': 'https://www.compassion.ch',
+    'license': 'AGPL-3',
+    'images': [],
+    'data': [
```
Remove empty Key.

ecino commented:
@MeetKD Thanks for your input, I added your suggestions.

open-net-sarl commented:
@ecino I have not tested your last modifications but the last time i have tested, it was OK.
We don't have plan using the module on version 10 but i will migrate this module for version 11.
But i would prefer having a fully merged module to begin the migration.

Thanks.

MeetKD commented:
Can you please check why travis is red?

ecino commented:
@MeetKD It was pylint issues in other modules. I fixed them.

ecino committed:
Add more tests to hr_attendance_management

dcoopennet commented:
@ecino Why is this PR taking so long to be merged.
We already did a version 11.0 of your module and we want to do a PR.
BUT if it is so slow to merge something we probably won't do this PR.

ecino commented:
@dcoopennet it's slow because people don't have time to review the PR... @MeetKD @pedrobaeza ?

leemannd commented:
@dcoopennet We are interested into a V11 of this PR/module.
We will do a review of both if you submit it. So they can be merged both quickly.

leemannd commented:
@dcoopennet To be merged, there is a need of review. @open-net-sarl  Did one, but there is nothing showed into the reviewers. You may want to do and and be visible in the "reviewer" section. It will help.

leemannd commented on hr_attendance_management/__init__.py:
```diff
@@ -0,0 +1,13 @@
+##############################################################################
+#
+#    Copyright (C) 2018 Compassion CH (http://www.compassion.ch)
+#    @author: Stephane Eicher <seicher@compassion.ch>
```
You should favor short header

grindtildeath replied:
BTW header is not absolutely required in `__init__.py` file. :wink:

leemannd commented on hr_attendance_management/__init__.py:
```diff
@@ -0,0 +1,13 @@
+
+from . import models
+from . import wizards
+from . import reports
```
missing newline EOF

leemannd commented on hr_attendance_management/data/annual_balance_cron.xml:
```diff
@@ -0,0 +1,27 @@
+            <field name="name">Compute annual extra hours balance</field>
+            <field name="interval_type">months</field>
+            <field name="interval_number">12</field>
+            <field name="nextcall" >2019-01-01 00:00:01</field>
```
I would have letted it empty and added a step into configuration in order to let the user know he has to configure it.

grindtildeath replied:
Simply remove it so default value is used :wink:

leemannd commented on hr_attendance_management/data/annual_balance_cron.xml:
```diff
@@ -0,0 +1,27 @@
+            <field name="active" eval="True"/>
+            <field name="interval_type">months</field>
+            <field name="interval_number">12</field>
+            <field name="nextcall" >2019-01-31 00:00:01</field>
```
I would have letted it empty and added a step into configuration in order to let the user know he has to configure it.

leemannd commented on hr_attendance_management/migrations/10.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,21 @@
+#    Releasing children from poverty in Jesus' name
+#    @author: Emanuel Cino <ecino@compassion.ch>
+#
+#    The licence is in the file __manifest__.py
```
Could you update to short license headers?

leemannd commented on hr_attendance_management/migrations/10.0.1.1.0/pre-migration.py:
```diff
@@ -0,0 +1,30 @@
+#    Releasing children from poverty in Jesus' name
+#    @author: Emanuel Cino <ecino@compassion.ch>
+#
+#    The licence is in the file __manifest__.py
```
Could you update to short license headers?

leemannd commented on hr_attendance_management/models/hr_attendance.py:
```diff
@@ -0,0 +1,106 @@
+
+    ##########################################################################
+    #                                 FIELDS                                 #
+    ##########################################################################
```
Can be dropped

leemannd commented on hr_attendance_management/models/hr_attendance.py:
```diff
@@ -0,0 +1,106 @@
+                                        string='Attendance day',
+                                        readonly=True)
+
+    due_hours = fields.Float(related='attendance_day_id.due_hours')
```
Do we want it readonly?

leemannd commented on hr_attendance_management/models/hr_attendance.py:
```diff
@@ -0,0 +1,106 @@
+
+    due_hours = fields.Float(related='attendance_day_id.due_hours')
+    total_attendance = fields.Float(
+        related='attendance_day_id.total_attendance')
```
Do we want it readonly?

leemannd commented on hr_attendance_management/models/hr_attendance.py:
```diff
@@ -0,0 +1,106 @@
+    total_attendance = fields.Float(
+        related='attendance_day_id.total_attendance')
+    has_change_day_request = fields.Boolean(
+        related='attendance_day_id.has_change_day_request')
```
Do we want it readonly?

leemannd commented on hr_attendance_management/models/hr_attendance.py:
```diff
@@ -0,0 +1,106 @@
+        """
+        attendance_days = self.env['hr.attendance.day']
+        for attendance in self:
+            date = attendance.check_in[:10]
```
Question, it's a datetime ?
Shouldn't we extract it with `fields.Datetime.from_string(date)` ?

grindtildeath replied:
If you want only the date out of it you can also do `fields.Date.from_string(attendance.check_in)`

leemannd commented on hr_holidays_imposed_days/models/hr_imposed_holidays.py:
```diff
@@ -113,10 +113,11 @@ def onchange_dates(self):
         return round(math.floor(diff))+1
 
-    @api.one
+    @api.multi
```
I don't know about this change.
`api.one` is officially deprecated, but still present into the official documentation when speaking about the `api.constrains`. https://www.odoo.com/documentation/10.0/reference/orm.html

leemannd reviewed (changes requested):
I didn't had the time to review all, but I have a few comments for now

grindtildeath commented:
@ecino Are you still planning to work on this ? 
@dcoopennet @open-net-sarl Both of you guys mentioned a v11.0 migration, do you plan to submit it to the OCA ? 

To get this PR merged :
 - Attend the comments by @leemannd 
 - Remove changes done to other modules, i.e. hr_employee_legacy_id,  hr_employee_seniority,  hr_expense_sequence,  hr_family,  hr_holidays_imposed_days,  hr_holidays_validity_date,  hr_recruitment_skill (most of those changes could be submitted in a separated PR if the goal is only to fix pylint)
 - Get more reviews (will come faster once those two points are done)

If you cannot do this, do you mind I submit a clean PR for this module in v10.0 so we can proceed with a v11.0 migration ?

ecino commented:
@grindtildeath Thanks for taking action on this. We have been improving the module in a separate repository because we are already using it in production. We have some improvements that we must push back into this branch and we will clean the pylint issues from other modules to let this PR be clean.

We plan to do this next week if that's ok for you? I would advise not to start something over this code base before we update it.

grindtildeath commented on hr_attendance_management/migrations/10.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,21 @@
+    # Recompute break duration with new field definition
+    cr.execute("""
+        UPDATE hr_attendance_break
+        SET additional_duration = total_duration - original_duration;
```
Is this needed ?

grindtildeath commented on hr_attendance_management/migrations/10.0.1.1.0/pre-migration.py:
```diff
@@ -0,0 +1,30 @@
+    # Rename module
+    openupgrade.update_module_names(
+        cr, [
+            ('hr_attendance_calendar', 'hr_attendance_management'),
```
Same here, I don't think this is needed.

grindtildeath commented on hr_attendance_management/models/hr_attendance_break.py:
```diff
@@ -0,0 +1,94 @@
+                stop = fields.Datetime.context_timestamp(
+                    rd.employee_id, stop)
+                stop = fields.Datetime.to_string(stop)[-8:]
+                rd.name = start + ' - ' + stop
```
:confused: Can you explain with a comment what you're doing here ? Using proper varnames would help to understand, and if you want to display only the start and stop times, you can use `strftime` with a proper format instead of slicing the string, and you won't need a comment to explicit what you're doing.

grindtildeath commented on hr_attendance_management/models/hr_attendance_break.py:
```diff
@@ -0,0 +1,94 @@
+    ##########################################################################
+    #                             FIELDS METHODS                             #
+    ##########################################################################
+    @api.multi
```
Add `api.depends`

grindtildeath commented on hr_attendance_management/models/hr_attendance_break.py:
```diff
@@ -0,0 +1,94 @@
+        for rd in self:
+            if rd.is_offered:
+                rd.name = 'Free break'
+            elif not rd.start:
```
If `start` is to be computed, I'm not sure this should happen ? (maybe should fix function dependency)

eicher31 replied:
Work as expected

grindtildeath commented on hr_attendance_management/models/hr_attendance_break.py:
```diff
@@ -0,0 +1,94 @@
+                rd.name = start + ' - ' + stop
+
+    @api.multi
+    @api.depends('previous_attendance', )
```
Add `next_attendance` ?

eicher31 replied:
It's not necessary because at each attendance write, all the attendance_break of the attendance_day are deleted and recreate. I know is not the best method to do... but I don't have the time to change it.

grindtildeath commented on hr_attendance_management/models/hr_attendance_break.py:
```diff
@@ -0,0 +1,94 @@
+    @api.multi
+    def _compute_system_modified(self):
+        for rd in self:
+            rd.system_modified = rd.additional_duration
```
Is this really needed ?

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+                                     'Attendances', readonly=True)
+
+    has_change_day_request = fields.Boolean(
+        compute='_compute_has_linked_change_day_request', store=True,
```
Would be great to rename compute method to `_compute_has_change_day_request` to be in line with field name.

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+    def _compute_has_linked_change_day_request(self):
+        for att_day in self:
+            res = att_day.get_related_forced_due_hours()
+            att_day.has_change_day_request = len(res) == 1
```
It feels weird that `has_change_day_request` would be False if `res > 1`.

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+                att_day.date).strftime('%A').title()
+
+    @api.multi
+    @api.depends('working_day')
```
Add `date`.

eicher31 replied:
working_day already depends of date

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+            balance = self.env.cr.fetchone()
+            balance = balance[0] if balance else 0
+
+            if balance > max_extra_hours > 0:
```
Is `> 0` required here ? :thinking:

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+    def write(self, vals):
+        super(HrAttendanceDay, self).write(vals)
+        if 'paid_hours' in vals or 'coefficient' in vals:
+            for att_day in self:
```
I feel like this loop might be unneeded here since there is another loop in `update_extra_hours_lost`. At least, `write`, then `search`, then `write`, to finally update a value is a lot to override `write`...

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+                ], order='date')
+                att_days_future.update_extra_hours_lost()
+
+        return True
```
Would be cleaner to return the result of the call to `super`

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+                })
+
+            att_break.write({
+                'system_modified': True,
```
This is a computed field, you shouldn't write on it...

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+                    'employee_id': att_day.employee_id.id,
+                    'attendance_day_id': att_day.id,
+                    'is_offered': True,
+                    'system_modified': True,
```
Again this is a computed field, so it's no use to set a value.

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+        week_day = att_date.weekday()
+        co_ids = self.env['hr.weekday.coefficient'].search([
+            ('day_of_week', '=', week_day)]).filtered(
+            lambda r: r.category_ids & rd.employee_id.category_ids)
```
Can't you move the conditions from `filtered` to `search` method ?

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+        :return: None
+        """
+        for att_day in self.filtered('attendance_ids'):
+            att_day.break_ids.unlink()
```
Wouldn't it be cleaner to `unlink` everything out of the loop ?
```
attendances = self.filtered('attendance_ids')
attendances.mapped('break_ids').unlink()
for att_day in attendances:
    ...
```

grindtildeath commented on hr_attendance_management/models/hr_attendance_day.py:
```diff
@@ -0,0 +1,487 @@
+                            'next_attendance': attendance.id,
+                        })
+                    previous_att = attendance
+                except StopIteration:
```
Isn't there any cleaner way to do this without `while True` and `try .. except` ?

grindtildeath commented on hr_attendance_management/models/hr_change_day_request.py:
```diff
@@ -0,0 +1,98 @@
+    def create(self, vals):
+        if 'user_id' not in vals:
+            manager = self.get_manager(vals['employee_id'])
+            vals['user_id'] = manager.user_id.id if manager else 1  # admin
```
IMO admin shouldn't be used as a default here, but a default should be defined as a parameter somewhere.

grindtildeath commented on hr_attendance_management/models/hr_change_day_request.py:
```diff
@@ -0,0 +1,98 @@
+    def unlink(self):
+        for request in self:
+            request.day1_id.unlink()
+            request.day2_id.unlink()
```
```
self.mapped('day1_id').unlink()
self.mapped('day2_id').unlink()
```

grindtildeath commented on hr_attendance_management/models/hr_change_day_request.py:
```diff
@@ -0,0 +1,98 @@
+            request.day1_id.unlink()
+            request.day2_id.unlink()
+
+        super(HrChangeDayRequest, self).unlink()
```
Add `return`

grindtildeath commented on hr_attendance_management/models/hr_change_day_request.py:
```diff
@@ -0,0 +1,98 @@
+    forced1 = fields.Float('Hours 1', related='day1_id.forced_due_hours')
+    forced2 = fields.Float('Hours 2', related='day2_id.forced_due_hours')
+
+    forced = fields.Char('Hours changed', compute='_compute_hours')
```
s/_compute_hours/_compute_forced to align field and function names.

grindtildeath commented on hr_attendance_management/models/hr_change_day_request.py:
```diff
@@ -0,0 +1,98 @@
+    @api.multi
+    def _compute_hours(self):
+        for h in self:
+            h.forced = h.forced2 if h.forced != 0 else h.forced1
```
How `forced` could be computed using `forced` ? :confused: 
Add `api.depends`

grindtildeath commented on hr_attendance_management/models/hr_employee.py:
```diff
@@ -0,0 +1,198 @@
+    @api.multi
+    @api.depends('attendance_days_ids.extra_hours', 'annual_balance')
+    def _compute_extra_hours(self, start_date=None, end_date=None,
+                             from_start_of_employment=True):
```
I don't think it's right to have these params on a `compute` method.

grindtildeath commented on hr_attendance_management/models/hr_employee.py:
```diff
@@ -0,0 +1,198 @@
+            # recreated
+            att_days = att_day.search(
+                [('date', '=', today), ('employee_id', '=', employee.id)])
+
```
Add here to spare the search, if there are no att_days
```
if att_days:
    continue
```

grindtildeath commented on hr_attendance_management/models/hr_employee.py:
```diff
@@ -0,0 +1,198 @@
+            # automatic computation of 01.01.XX 00:00:01.
+            # This will put an unexpected value in compute_extra_hours. It
+            # will be recomputed when employee.annual_balance will be set
+            employees._compute_extra_hours(start_date=start_previous_year,
```
:confused: Is there any good reason to call this explicitly since the values will be recomputed after the following update of `annual_balance` ?

grindtildeath commented on hr_attendance_management/models/hr_employee.py:
```diff
@@ -0,0 +1,198 @@
+        for employee in self:
+            if employee.extra_hours < 0:
+                employee.time_warning_balance = 'red'
+            elif employee.extra_hours >= 19:
```
How is this value defined ? No parameter ?

grindtildeath commented on hr_attendance_management/models/hr_employee.py:
```diff
@@ -0,0 +1,198 @@
+        for employee in self:
+            employee.extra_hours_today = \
+                '-' if float(employee.today_hour) < 0 else ''
+            employee.extra_hours_today += employee. \
```
Please compute the result first and do one update instead of two...

grindtildeath commented on hr_attendance_management/models/hr_employee.py:
```diff
@@ -0,0 +1,198 @@
+        for employee in self:
+            employee.extra_hours_formatted = \
+                '-' if employee.extra_hours < 0 else ''
+            employee.extra_hours_formatted += \
```
Please compute the result first and do one update instead of two...

grindtildeath commented on hr_attendance_management/models/hr_employee.py:
```diff
@@ -0,0 +1,198 @@
+            today_hour = employee.compute_today_hour()
+            thf = '-' if today_hour < 0 else ''
+            thf += employee.convert_hour_to_time(float(today_hour))
+            employee.today_hour_formatted = thf
```
This is exactly what I'm asking above :+1: :rofl:

grindtildeath commented on hr_attendance_management/models/hr_employee.py:
```diff
@@ -0,0 +1,198 @@
+        return '{:02d}:{:02d}'.format(*divmod(int(abs(float(hour) * 60)), 60))
+
+    @api.multi
+    def compute_today_hour(self):
```
Please call this `calc_*` to differentiate it from fields computing methods

grindtildeath commented on hr_attendance_management/models/hr_forced_due_hours.py:
```diff
@@ -0,0 +1,53 @@
+
+    @api.multi
+    def write(self, vals):
+        for record in self:
```
Looping inside a write is not a good idea.

eicher31 replied:
Ok... could you explain me the reasons?

grindtildeath replied:
Performance will be very bad when writing on multiple records as best explained by Odoo. 
Have a look at https://www.youtube.com/watch?v=dQYGau81RYg (slides here : https://www.odoo.com/slides/slide/orm-performance-optimizations-best-practices-506).
Part two is what applies here because we're already in a O(n) for 'simple' writes and with the successive calls to `recompute_due_hours` we'll end up having lots of processing and I won't calculate here how many requests this would imply. :wink:

grindtildeath commented on hr_attendance_management/models/hr_weekday_coefficient.py:
```diff
@@ -0,0 +1,44 @@
+        ('4', 'Friday'),
+        ('5', 'Saturday'),
+        ('6', 'Sunday'),
+        ('7', 'Default')], 'Day of Week', required=True, index=True,
```
I didn't know `('7', 'Default')` is a day of week :confounded: :fearful:

yvaucher replied:
@ecino I'm a bit out of context but a `help` tooltip could be useful to explain what 'Default' is.

grindtildeath reviewed (changes requested):
It took me more than 2 hours to review and I didn't check views and the wizard yet, but I have the feeling there are quite a few design issues with too many recomputations.
I also have the feeling it's really complex for things which could probably be done more simply. :pensive: 
Anyway I'll wait for your changes next week before coming back...

ecino commented:
Thanks for your review @grindtildeath! You're right about complexity and too many recomputations, it's something we thought as well. We simplified a little bit since then but there is still room for improvements. If you would like to contribute as well, you are welcome to do so. Again, we just need first to update this PR to our latest version. We'll keep you posted.

eicher31 committed:
Improve due hours during leave

Increase the accurate of due hours during a leave.
Keep the history of working schedule in the object hr.attendance.

eicher31 commented:
@grindtildeath we added our last modifications and fix some of your propositions, but not all because we don't have enough time now. feel free to contribute if you wish so.

jo-bloiz-ornot committed:
[ADD] new module hr_attendance_management
work by CompassionCH

jo-bloiz-ornot committed:
[UPD] module hr_attendance_management
Correction of failed checks

c-meier committed:
[FIX] Extra hour now waits for validation.

jo-bloiz-ornot committed:
CO-2442 test HR

Rewrite test hr content
add missing values for using demo correclty

[FIX] travis error on module hr_attendance_management

[FIX] travis error on module hr_attendance_management

[FIX] travis error on module hr_attendance_management

[FIX] travis error on module hr_attendance_management

vwvw committed:
CO-2505 annualisation of annual hours and refactoring of extra hours
computations

The overtime computation was moved from a past-dependant computation in
each attendance_day to a continuous computation in hr_employee. This
removes dependencies on past data apart from the yearly balances.

vwvw committed:
Addressing @ecino comments and removing unnecessary cron

vwvw committed:
Change name of variables

ecino committed:
Merge pull request #13 from vwvw/CO-2505

CO-2505 annualisation of annual hours and refactoring of extra hours computations

vwvw committed:
Modifying views

ecino committed:
Merge pull request #14 from vwvw/CO-2505

Modifying views

vwvw committed:
Fix for null balance after migration

vwvw committed:
Fix as requested by David

ecino committed:
Merge pull request #17 from CompassionCH/new-annual-limit

New annual limit

ecino committed:
CO-2608 FIX CRON create attendance day that was taking possibly old date

Theyiot committed:
CO-2610 Minor modifications and corrections of hr_attendance views

Theyiot committed:
CO-2610 Put back field extra_hours_lost, since ot was not explicitely asked to be removed

Theyiot committed:
CO-2610 improvement according to feedback

ecino committed:
Merge pull request #18 from Theyiot/theo

CO-2610: Corrections in hr_attendance views

vwvw committed:
CO-2622 Automatically recompute old balances (#19)

* CO-2622 Automatically recompute old balances

* Fix after review

samybucher committed:
CO-2669 Remove read_only on field working_schedule_id.

quentingigon committed:
old records of balance.evolution.graph are now deleted before graph display

quentingigon committed:
tmp commit to save state before changing machine

quentingigon committed:
new system for attendance management

quentingigon committed:
add all

quentingigon committed:
changed the was the history is built and updated to include continuous_cap

quentingigon committed:
view for hitoriy entries is done and the periods are now updated when the continuous_cap is changed

quentingigon committed:
removed all traces of penultimate and co. added beginning date to settings. fixes and cleanup

quentingigon committed:
comments, bug fixes

quentingigon committed:
added migrations for balance computation

quentingigon committed:
corrected migration

quentingigon committed:
firt batch of tests working and corrections to the code

quentingigon committed:
more fixes and update to old tests. there are still some tests that dont pass

quentingigon committed:
corrections. there are still some bugs with balance and lost_hours computation i think

quentingigon committed:
added security rules for periods and some comments for tests

quentingigon committed:
added tests and made corrections to models

quentingigon committed:
changed post-migration to avoid setting a new value for balance and correction for a bug during period creation

quentingigon committed:
correction of annual_balance computation and ignoring file permissions changes

quentingigon committed:
Revert "add all"

This reverts commit 0462285f340c3ee9db789681d300381189b20443.

quentingigon committed:
corrections for the migration, it should now create a period for 2018 with correct values and compute balance then

quentingigon committed:
changed period bound logic to make the end_date exclusive to comply with pre-existing methodology

quentingigon committed:
correction for the view of initial balance

quentingigon committed:
updates to migration and period creation. now cap is not taken into account for period creation

quentingigon committed:
Move logic inside periods and migration changes

quentingigon committed:
fixed balance computation and period creation during migration

quentingigon committed:
we can now create a period just after an existing one + get my total balance is now working

quentingigon committed:
more coherence in balance computation and surrounding_periods work

quentingigon committed:
now if a period is inside a new period, deltes it

quentingigon committed:
the modification of the balance or the final_balance of the previous period now calls a recompute of all successive periods

quentingigon committed:
added a new action to create period from the employee view

quentingigon committed:
balance is recomputed when initial_balance is modified but there is an error ocurring that does not affect the result

quentingigon committed:
periods now use start_date and end_date chosen by the user as both inclusive bounds

quentingigon committed:
fix migration

quentingigon committed:
fix for recursive error when changing initial_balance

quentingigon committed:
fix for bug when creating new period with old previous period overlapping

quentingigon committed:
refactor and comments for hr.employee.period

quentingigon committed:
fix migration

quentingigon committed:
fix for bad balance computation when we have data for the current day

quentingigon committed:
bug fixing and value testing

quentingigon committed:
applied changes request

quentingigon committed:
updated and fixed tests for periods and balance computation

quentingigon committed:
finished applying change requests

quentingigon committed:
removed store=True for balance in employee

quentingigon committed:
balance is stored again

quentingigon committed:
fix for test_att_day, now need to fix tests on rules

quentingigon committed:
fixed all tests

quentingigon committed:
Mini fix

quentingigon committed:
added tests and fixed some bugs/tests

quentingigon committed:
fix for recursion error

quentingigon committed:
fixed tests and bug fixing

quentingigon committed:
added tests and fixed non-update of balance when modifing an attendance

eicher31 committed:
Merge pull request #22 from quentingigon/annual_balance_computation_fix

CO-2728 balance computation changes

quentingigon committed:
fix flake errors

quentingigon committed:
fix tests and stuff

quentingigon committed:
restricted view of initial balance to manager only

eicher31 committed:
Add CODEOWNER file

eicher31 committed:
Merge pull request #23 from quentingigon/annual_balance_computation_fix

CO-2728 Annual balance computation fix

quentingigon committed:
added label of initial_balance to hr manager group

eicher31 committed:
Merge pull request #25 from quentingigon/annual_balance_computation_fix

fix for hr_attendance

ecino commented:
This module is becoming huge and needs better reviews. We will split it into different components so that each one can be better reviewed/tested.

leemannd commented:
@ecino Don't hesitate to ping me for a review / functional testing

ecino commented:
@leemannd Thanks, I will do it!
